crates/ruff/src/rules/ruff/rules/unnecessary_iterable_allocation_for_first_element.rs

```diff
@@ -0,0 +1,223 @@
+use num_traits::ToPrimitive;
+use ruff_diagnostics::{AlwaysAutofixableViolation, Diagnostic, Edit, Fix};
+use ruff_macros::{derive_message_formats, violation};
+use rustpython_parser::ast::{self, Comprehension, Constant, Expr};
+
+use crate::checkers::ast::Checker;
+use crate::registry::AsRule;
+
+/// ## What it does
+/// Ensures that instead of creating a new list and indexing into it to find the first element of a
+/// collection (e.g., `list(...)[0]`), Python iterators are used.
+///
+/// ## Why is this bad?
+/// Creating a new list of great size can involve significant memory/speed concerns. Python's `next(iter(...))`
+/// pattern can be used in lieu of creating a new list. This pattern will lazily fetch the first
+/// element of the collection, avoiding the memory overhead involved with new list allocation.
+/// `next(iter(...))` also is much faster since the list itself doesn't get initialized at once.
+///
+/// ## Example
+/// ```python
+/// x = range(1000000000000)
+/// return list(x)[0]
+/// ```
+///
+/// Use instead:
+/// ```python
+/// x = range(1000000000000)
+/// return next(iter(x))
+/// ```
+///
+/// ## References
+/// - [Iterators and Iterables in Python: Run Efficient
+/// Iterations](https://realpython.com/python-iterators-iterables/#when-to-use-an-iterator-in-python)
+#[violation]
+pub(crate) struct UnnecessaryIterableAllocationForFirstElement {
+    arg: String,
+}
+
+impl UnnecessaryIterableAllocationForFirstElement {
+    pub(crate) fn new(arg: String) -> Self {
+        Self { arg }
+    }
+}
+
+impl AlwaysAutofixableViolation for UnnecessaryIterableAllocationForFirstElement {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!(
+            "Prefer `next(iter({}))` over `list({})[0]` or equivalent list comprehension",
+            self.arg, self.arg
+        )
+    }
+
+    fn autofix_title(&self) -> String {
+        format!("Replace with `next(iter({}))`", self.arg)
+    }
+}
```

Comment:
Oh, sorry I missed this in my previous review but for slice the suggestion will be different.

```
Replace with `[next(iter({}))]`
```

Similar change in the violation message.

Reply:
Oh, good point. Thanks!